client/src/types/snapshot.rs

```diff
@@ -212,24 +215,24 @@ impl Snapshot {
     /// TODO: Add associated data.
     pub fn write_to_snapshot(&self, snapshot_path: &SnapshotPath, use_key: UseKey) -> Result<(), SnapshotError> {
         let state = self.get_snapshot_state()?;
-        let data = bincode::serialize(&state)?;
+        let data = Zeroizing::new(bincode::serialize(&state)?);
 
-        let key = match use_key {
-            UseKey::Key(k) => k,
+        match use_key {
+            UseKey::Key(k) => snapshot::encrypt_file(&data, snapshot_path.as_path(), &k, &[]).map_err(|e| e.into()),
             UseKey::Stored(loc) => {
                 let (vid, rid) = loc.resolve();
                 let pkey = self.keystore.get_key(vid).ok_or(SnapshotError::SnapshotKey(vid, rid))?;
-                let mut data = Vec::new();
-                self.db.get_guard::<Infallible, _>(&pkey, vid, rid, |guarded_data| {
-                    let guarded_data = guarded_data.borrow();
-                    data.extend_from_slice(&guarded_data);
-                    Ok(())
-                })?;
-                data.try_into().map_err(|_| SnapshotError::SnapshotKey(vid, rid))?
+                Ok(self
+                    .db
+                    .get_guard::<_, SnapshotError, _>(&pkey, vid, rid, |guarded_data| {
+                        let key_ref = guarded_data.borrow();
+                        let key: &snapshot::Key = (*key_ref)
+                            .try_into()
+                            .map_err(|_| SnapshotError::SnapshotKey(vid, rid))?;
+                        Ok(snapshot::encrypt_file(&data, snapshot_path.as_path(), key, &[])?)
+                    })?)
```

Comment:
Can the `Ok(..?)` not be removed?
```suggestion
                self
                    .db
                    .get_guard::<_, SnapshotError, _>(&pkey, vid, rid, |guarded_data| {
                        let key_ref = guarded_data.borrow();
                        let key: &snapshot::Key = (*key_ref)
                            .try_into()
                            .map_err(|_| SnapshotError::SnapshotKey(vid, rid))?;
                        snapshot::encrypt_file(&data, snapshot_path.as_path(), key, &[])
                    })
```

Reply:
This trick was used to easily convert between error types, rust does it automatically with From trait. Not sure if it's required here. Otherwise we'd have to use map_err explicitly.

Reply:
If it returns different errors then it's fine to do this for the conversion :+1:

Reply:
No, it's required here.